File: app/src/main/java/com/revotechs/calculator/dao/HistoryDao.java
package com.revotechs.calculator.dao;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.revotechs.calculator.tools.DBHelper;
import com.revotechs.calculator.tools.HistoryItem;

import java.util.ArrayList;
import java.util.List;

/**
 * Project Calculator
 *
 * Created on 30.05.2017
 * @author CriticalGnome
 */

public class HistoryDao {

    private static final String TABLE_HISTORY = "history";
    private static final String FIELD_ID = "id";
    private static final String FIELD_COMMENT = "comment";
    private static final String FIELD_DATE = "date";
    private static final String FIELD_EXPRESSION = "expression";
    private static final String FIELD_RESULT = "result";
    private static final String ORDERING_DESC = "DESC";
    private static final String HISTORY_DAO = "historyDAO";
    private static final String NO_DATA = "No data";
    private static final String UPDATED = " updated";
    private static final String DELETED = " deleted";
    private static final String EQUAL_SIGN = " = ";
    private static volatile HistoryDao instance;
    private DBHelper dbHelper;
    private ContentValues cv;
    private SQLiteDatabase db;

    private HistoryDao() {}

    public static HistoryDao getInstance() {
        if (instance == null) {
            synchronized (HistoryDao.class) {
                if (instance == null) {
                    instance = new HistoryDao();
                }
            }
        }
        return instance;
    }

    public Long create(HistoryItem item, Context context) {
        dbHelper = new DBHelper(context);
        db = dbHelper.getWritableDatabase();
        cv = new ContentValues();
        cv.put(FIELD_DATE, item.getDate());
        cv.put(FIELD_EXPRESSION, item.getExpression());
        cv.put(FIELD_RESULT, item.getResult());
        cv.put(FIELD_COMMENT, item.getComment());
        Long id = db.insert(TABLE_HISTORY, null, cv);
        item.setId(id);
        Log.d(HISTORY_DAO, item.toString());
        dbHelper.close();
        return id;
    }

    public HistoryItem getOne(Long id, Context context) {
        dbHelper = new DBHelper(context);
        db = dbHelper.getReadableDatabase();
        HistoryItem item = new HistoryItem();
        Cursor cursor = db.query(TABLE_HISTORY, null, FIELD_ID + " = ?", new String[] {id.toString()}, null, null, null);
        if (cursor.moveToFirst()) {
            int idColIndex = cursor.getColumnIndex(FIELD_ID);
            int dateColIndex = cursor.getColumnIndex(FIELD_DATE);
            int expressionColIndex = cursor.getColumnIndex(FIELD_EXPRESSION);
            int resultColIndex = cursor.getColumnIndex(FIELD_RESULT);
            int commentColIndex = cursor.getColumnIndex(FIELD_COMMENT);
            do {
                item.setId(cursor.getLong(idColIndex));
                item.setDate(cursor.getString(dateColIndex));
                item.setExpression(cursor.getString(expressionColIndex));
                item.setResult(cursor.getString(resultColIndex));
                item.setComment(cursor.getString(commentColIndex));
            } while (cursor.moveToNext());
        } else {
            Log.d(HISTORY_DAO, NO_DATA);
        }
        Log.d(HISTORY_DAO, item.toString());
        cursor.close();
        dbHelper.close();
        return item;
    }

    public List<HistoryItem> getAll(Context context) {
        dbHelper = new DBHelper(context);
        db = dbHelper.getReadableDatabase();
        List<HistoryItem> items = new ArrayList<>();
        Cursor cursor = db.query(TABLE_HISTORY, null, null, null, null, null, FIELD_ID + " " + ORDERING_DESC);
        if (cursor.moveToFirst()) {
            int idColIndex = cursor.getColumnIndex(FIELD_ID);
            int dateColIndex = cursor.getColumnIndex(FIELD_DATE);
            int expressionColIndex = cursor.getColumnIndex(FIELD_EXPRESSION);
            int resultColIndex = cursor.getColumnIndex(FIELD_RESULT);
            int commentColIndex = cursor.getColumnIndex(FIELD_COMMENT);
            do {
                HistoryItem item = new HistoryItem();
                item.setId(cursor.getLong(idColIndex));
                item.setDate(cursor.getString(dateColIndex));
                item.setExpression(cursor.getString(expressionColIndex));
                item.setResult(cursor.getString(resultColIndex));
                item.setComment(cursor.getString(commentColIndex));
                items.add(item);
            } while (cursor.moveToNext());
        } else {
            Log.d(HISTORY_DAO, NO_DATA);
        }
        Log.d(HISTORY_DAO, items.toString());
        cursor.close();
        dbHelper.close();
        return items;
    }

    public void update(HistoryItem item, Context context) {
        dbHelper = new DBHelper(context);
        db = dbHelper.getWritableDatabase();
        cv = new ContentValues();
        cv.put(FIELD_DATE, item.getDate());
        cv.put(FIELD_EXPRESSION, item.getExpression());
        cv.put(FIELD_RESULT, item.getResult());
        cv.put(FIELD_COMMENT, item.getComment());
        String where = FIELD_ID + EQUAL_SIGN + item.getId();
        db.update(TABLE_HISTORY, cv, where, null);
        Log.d(HISTORY_DAO, FIELD_ID + EQUAL_SIGN +  item.getId() + UPDATED);
    }

    public void delete(Long id, Context context) {
        dbHelper = new DBHelper(context);
        db = dbHelper.getWritableDatabase();
        db.delete(TABLE_HISTORY, FIELD_ID + EQUAL_SIGN + id, null);
        Log.d(HISTORY_DAO, FIELD_ID + EQUAL_SIGN +  id + DELETED);
    }

}

File: app/src/main/java/com/revotechs/calculator/dao/HistoryDao.kt
package com.revotechs.calculator.dao

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.util.Log
import com.revotechs.calculator.entitiy.HistoryItem
import com.revotechs.calculator.tool.DBHelper
import java.util.*

/**
 * Project Calculator
 * Created on 30.05.2017
 * @author CriticalGnome
 */
class HistoryDao {

    fun create(item: HistoryItem, context: Context): Long {
        val dbHelper = DBHelper(context)
        val db = dbHelper.writableDatabase
        val cv = ContentValues()
        cv.put(FIELD_DATE, item.date)
        cv.put(FIELD_EXPRESSION, item.expression)
        cv.put(FIELD_RESULT, item.result)
        cv.put(FIELD_COMMENT, item.comment)
        if (item.locked) {
            cv.put(FIELD_LOCKED, 1)
        } else {
            cv.put(FIELD_LOCKED, 0)
        }
        val id = db.insert(TABLE_HISTORY, null, cv)
        item.id = id
        Log.d(HISTORY_DAO, item.toString())
        dbHelper.close()
        return id
    }

//    fun getOne(id: Long, context: Context): HistoryItem {
//        val dbHelper = DBHelper(context)
//        val db = dbHelper.readableDatabase
//        var item = HistoryItem(0L, "", "", "", "", false)
//        val cursor = db.query(TABLE_HISTORY, null, "$FIELD_ID = ?", arrayOf(id.toString()), null, null, null)
//        if (cursor.moveToFirst()) {
//            do {
//                item = constructItem(cursor)
//            } while (cursor.moveToNext())
//        } else {
//            Log.d(HISTORY_DAO, NO_DATA)
//        }
//        Log.d(HISTORY_DAO, item.toString())
//        cursor.close()
//        dbHelper.close()
//        return item
//    }

    fun getAll(context: Context): List<HistoryItem> {
        val dbHelper = DBHelper(context)
        val db = dbHelper.readableDatabase
        val items = ArrayList<HistoryItem>()
        val cursor = db.query(TABLE_HISTORY, null, null, null, null, null, "$FIELD_ID DESC")
        if (cursor.moveToFirst()) {
            do {
                items.add(constructItem(cursor))
            } while (cursor.moveToNext())
        } else {
            Log.d(HISTORY_DAO, NO_DATA)
        }
        Log.d(HISTORY_DAO, items.toString())
        cursor.close()
        dbHelper.close()
        return items
    }

    fun update(item: HistoryItem, context: Context) {
        val dbHelper = DBHelper(context)
        val db = dbHelper.writableDatabase
        val cv = ContentValues()
        cv.put(FIELD_DATE, item.date)
        cv.put(FIELD_EXPRESSION, item.expression)
        cv.put(FIELD_RESULT, item.result)
        cv.put(FIELD_COMMENT, item.comment)
        if (item.locked) {
            cv.put(FIELD_LOCKED, 1)
        } else {
            cv.put(FIELD_LOCKED, 0)
        }
        db.update(TABLE_HISTORY, cv, "$FIELD_ID = ${item.id}", null)
        Log.d(HISTORY_DAO, "$FIELD_ID = ${item.id} updated")
    }

    fun delete(id: Long, context: Context) {
        val dbHelper = DBHelper(context)
        val db = dbHelper.writableDatabase
        db.delete(TABLE_HISTORY, "$FIELD_ID = $id", null)
        Log.d(HISTORY_DAO, "$FIELD_ID = $id deleted")
    }

    fun search(search: String, context: Context): List<HistoryItem> {
        val dbHelper = DBHelper(context)
        val db = dbHelper.readableDatabase
        val items = ArrayList<HistoryItem>()
        val cursor = db.query(
                TABLE_HISTORY,
                null,
                "$FIELD_DATE LIKE '%$search%' or $FIELD_EXPRESSION LIKE '%$search%' or $FIELD_RESULT LIKE '%$search%' or $FIELD_COMMENT LIKE '%$search%'",
                null,
                null,
                null,
                "$FIELD_ID DESC")
        if (cursor.moveToFirst()) {
            do {
                items.add(constructItem(cursor))
            } while (cursor.moveToNext())
            Log.d(HISTORY_DAO, items.toString())
        } else {
            Log.d(HISTORY_DAO, NO_DATA)
        }
        cursor.close()
        dbHelper.close()
        return items
    }

    fun clearHistory(context: Context) {
        val dbHelper = DBHelper(context)
        val db = dbHelper.writableDatabase
        db.delete(TABLE_HISTORY, "$FIELD_LOCKED = 0", null)
    }

    private fun constructItem(cursor: Cursor): HistoryItem {
        return HistoryItem(
                cursor.getLong(cursor.getColumnIndex(FIELD_ID)),
                cursor.getString(cursor.getColumnIndex(FIELD_DATE)),
                cursor.getString(cursor.getColumnIndex(FIELD_EXPRESSION)),
                cursor.getString(cursor.getColumnIndex(FIELD_RESULT)),
                cursor.getString(cursor.getColumnIndex(FIELD_COMMENT)),
                cursor.getInt(cursor.getColumnIndex(FIELD_LOCKED)) == 1)
    }

    companion object {

        private const val HISTORY_DAO = "historyDAO"
        private const val TABLE_HISTORY = "history"
        private const val FIELD_ID = "id"
        private const val FIELD_DATE = "date"
        private const val FIELD_EXPRESSION = "expression"
        private const val FIELD_RESULT = "result"
        private const val FIELD_COMMENT = "comment"
        private const val FIELD_LOCKED = "locked"
        private const val NO_DATA = "No data"

    }
}
